File: python/OOP/pattern_singleton.py

'''
Singleton Pattern ensures only one instance of a struct exists.

Useful for:
- database connections
- logging systems
- configuration settings

✅ Thread-safe using sync.Once
✅ Ensures a single instance
'''

class Singleton:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

obj1 = Singleton()
obj2 = Singleton()
print(obj1 is obj2)  # Output: True (both variables reference the same instance)

File: golang/07_application_structure_and_patterns/pattern_singleton.go
package main
import (
    "fmt"
    "sync"
)

/*
Singleton Pattern ensures only one instance of a struct exists.

Useful for:
- database connections
- logging systems
- configuration settings

✅ Thread-safe using sync.Once
✅ Ensures a single instance
*/

type Logger struct{}

var instance *Logger
var once sync.Once

func GetLogger() *Logger {
    once.Do(func() {
        instance = &Logger{}
    })
    return instance
}

func (l *Logger) Log(msg string) {
    fmt.Println(msg)
}

func singleton_pattern() {
    logger1 := GetLogger()
    logger2 := GetLogger()

    logger1.Log("Hello, Singleton!")

    fmt.Println(logger1 == logger2) // true, same instance
}
